### PR TITLE
HADOOP-17020. Improving RawFileSystem Performance

### PR DESCRIPTION
**Pull request by mehakmeet**
Contributed by: Mehakmeet Singh

Tested by: mvn -T 1C -Dparallel-tests -DtestsThreadCount=8 clean verify
Region: East US, West US

### CODE REVIEW

**Comment by mehakmeet** [comment]
Tests not required, due to existing tests validating the patch.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  28m 15s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  26m 20s |  trunk passed  |
| +1 :green_heart: |  compile  |  22m 49s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   1m  3s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 42s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m 32s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  2s |  trunk passed  |
| +0 :ok: |  spotbugs  |   2m 47s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 44s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  2s |  the patch passed  |
| +1 :green_heart: |  compile  |  22m 14s |  the patch passed  |
| +1 :green_heart: |  javac  |  22m 14s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m  0s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 41s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 18s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  5s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   2m 38s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  10m  1s |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 52s |  The patch does not generate ASF License warnings.  |
|  |   | 161m 21s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2063/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2063 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux f5a67e909337 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a8610c15c49 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2063/1/testReport/ |
| Max. process+thread count | 1631 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2063/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
relates to #2002

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java [hunk 1]
```diff
@@ -100,6 +101,7 @@ public File pathToFile(Path path) {
   public void initialize(URI uri, Configuration conf) throws IOException {
     super.initialize(uri, conf);
     setConf(conf);
+    defaultBlockSize = getDefaultBlockSize(new Path("."));
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java:104]: must be absolute fs uri

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java [hunk 2]
```diff
@@ -518,7 +520,12 @@ public boolean delete(Path p, boolean recursive) throws IOException {
     }
     return new FileStatus[] {
         new DeprecatedRawLocalFileStatus(localf,
-        getDefaultBlockSize(f), this) };
+        defaultBlockSize, this) };
+  }
+
+  @Override
+  public boolean exists(Path f) throws IOException {
+    return pathToFile(f).exists();
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java:528]: is this a new method, or just an accidental patch quirk

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  27m 28s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  19m 22s |  trunk passed  |
| +1 :green_heart: |  compile  |  17m  4s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 29s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 35s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  6s |  trunk passed  |
| +0 :ok: |  spotbugs  |   2m  9s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m  5s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 20s |  the patch passed  |
| +1 :green_heart: |  javac  |  16m 21s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 27s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m  7s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  4s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   2m 16s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 24s |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 54s |  The patch does not generate ASF License warnings.  |
|  |   | 134m 45s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2063/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2063 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux a41f55acb385 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 93b121a9717 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2063/2/testReport/ |
| Max. process+thread count | 1374 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2063/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
Looks good. One question though: Is the change on L520 Meant to be part of this patch? I can see it is an optimisation -But it is not listed in the JIRA

If you want it in, we need to change it's title to, say "improve RawFS performance" and list the changes in the commit message and JIRA

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java (hunk 2 above)
Comment (mehakmeet) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java:528]: This is a new method(Overriden) for this patch.

**Pull request merged**